### PR TITLE
Add support for reading the dynamic symbol table from PT_DYNAMIC

### PR DESCRIPTION
**Pull request by clayborg**
Allow LLDB to parse the dynamic symbol table from an ELF file or memory image in an ELF file that has no section headers. This patch uses the ability to parse the PT_DYNAMIC segment and find the DT_SYMTAB, DT_SYMENT, DT_HASH or DT_GNU_HASH to find and parse the dynamic symbol table if the section headers are not present. It also adds a helper function to read data from a .dynamic key/value pair entry correctly from the file or from memory.

### CODE REVIEW

File: lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp [hunk 1]
```diff
@@ -3895,3 +3924,103 @@ std::optional<lldb_private::DataExtractor> ObjectFileELF::GetDynamicData() {
...
+std::optional<DataExtractor>
+ObjectFileELF::GetDynsymDataFromDynamic(uint32_t &num_symbols) {
+  // Every ELF file which represents an executable or shared library has
+  // mandatory .dynamic entries. The DT_SYMTAB value contains a pointer to the
+  // symbol table, and DT_SYMENT contains the size of a symbol table entry.
+  // We then can use either the DT_HASH or DT_GNU_HASH to find the number of
+  // symbols in the symbol table as the symbol count is not stored in the
+  // .dynamic section as a key/value pair.
+  //
+  // When loading and ELF file from memory, only the program headers end up
```
Comment (Jlalond) [inline lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:3938]: Great comment, but I would rephrase that only the program headers are guaranteed to be mapped

File: lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp [hunk 2]
```diff
@@ -3895,3 +3924,103 @@ std::optional<lldb_private::DataExtractor> ObjectFileELF::GetDynamicData() {
...
+  lldb::offset_t offset;
+  if (hash) {
+    // The DT_HASH header contains the number of symbols in the "nchain"
+    // member. The header looks like this:
+    // struct DT_HASH_HEADER {
+    //   uint32_t nbucket;
+    //   uint32_t nchain;
+    // };
+    if (auto data = ReadDataFromDynamic(hash, 8)) {
+      offset = 4;
```
Comment (Jlalond) [inline lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:3969]: Should we make a comment why we're ignoring the first field

**Review by Jlalond — APPROVED** [review]
I left some comments. All code layout options and comment concerns, otherwise LGTM!

File: lldb/test/Shell/ObjectFile/ELF/elf-dynsym.yaml [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+# CHECK: [ 1] 2 X Undefined 0x0000000000000000 0x0000000000000000 0x00000012 puts
+# CHECK: [ 2] 3 Undefined 0x0000000000000000 0x0000000000000000 0x00000020 _ITM_deregisterTMCloneTable
+# CHECK: [ 3] 4 Undefined 0x0000000000000000 0x0000000000000000 0x00000020 __gmon_start__
+# CHECK: [ 4] 5 Undefined 0x0000000000000000 0x0000000000000000 0x00000020 _ITM_registerTMCloneTable
+# CHECK: [ 5] 6 X Code 0x0000000000001135 0x0000000000000016 0x00000012 baz()
+# CHECK: [ 6] 7 X Code 0x000000000000111f 0x0000000000000016 0x00000012 bar()
+# CHECK: [ 7] 8 X Code 0x000000000000114b 0x0000000000000016 0x00000012 biz()
+# CHECK: [ 8] 9 X Code 0x0000000000001109 0x0000000000000016 0x00000012 foo()
+
+--- !ELF
```
Comment (labath) [inline lldb/test/Shell/ObjectFile/ELF/elf-dynsym.yaml:37]: This input contains a lot of data that's irrelevant for the test. Can you make it smaller? For example, something like `llvm-mc -triple=x86_64-pc-linux -filetype=obj -o - - <<<".globl defined, undefined; defined:" | ld.lld /dev/stdin -o - --hash-style=gnu -export-dynamic` produces a much smaller file which still has a dynamic section, and you can control exactly which entries it contains (in practice you may want to expand the input slightly, just so that all the symbols don't end up with the same value).

You could yamlize that, or even just leave that as the input to the test itself (it should be deterministic enough). If you go with the second option, you can add `-z nosectionheader` to the linker command to skip the `llvm-strip` step.

I think think that another test with the DT_HASH hash style would be in order (just change the above command to --hash-style=sysv)

Comment (clayborg) [reply lldb/test/Shell/ObjectFile/ELF/elf-dynsym.yaml:37]: I am happy to use the llvm-mc + ld.lld trick, bit I need:
- the output file to have a .dynamic section or PT_DYNAMIC (above example doesn't)
- it needs to have a dynamic symbol table (SHT_DYNSYM or .dynamic section needs to have `DT_SYMTAB` and `DT_SYMENT` entries
- Needs to have a DT_HASH or DT_GNU_HASH

I don't know much about the `llvm-mc` tool, but if you can get it to generate a file that i can use I will use it.

Comment (labath) [reply lldb/test/Shell/ObjectFile/ELF/elf-dynsym.yaml:37]: I see the problem. I used the version of lld on my system, and it looks like the behavior has changed slightly for the version at HEAD. Sorry about that. I believe this has everything you need:
```
$ bin/llvm-mc -triple=x86_64-pc-linux -filetype=obj -o - - <<<".globl defined, undefined; defined:" | bin/ld.lld /dev/stdin -o - --hash-style=gnu -export-dynamic -shared -z nosectionheader  | llvm-readelf - -d -l

Elf file type is DYN (Shared object file)
Entry point 0x0
There are 6 program headers, starting at offset 64

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000000000040 0x0000000000000040 0x000188 0x000188 R   0x8
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000243 0x000243 R   0x1000
  LOAD           0x000248 0x0000000000002248 0x0000000000002248 0x000060 0x000db8 RW  0x1000
  DYNAMIC        0x000248 0x0000000000002248 0x0000000000002248 0x000060 0x000060 RW  0x8
  GNU_RELRO      0x000248 0x0000000000002248 0x0000000000002248 0x000060 0x000db8 R   0x1
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x0

 Section to Segment mapping:
  Segment Sections...
   00     
   01     
   02     
   03     
   04     
   05     
Dynamic section at offset 0x248 contains 6 entries:
  Tag                Type       Name/Value
  0x0000000000000006 (SYMTAB)   0x1c8
  0x000000000000000b (SYMENT)   24 (bytes)
  0x0000000000000005 (STRTAB)   0x230
  0x000000000000000a (STRSZ)    19 (bytes)
  0x000000006ffffef5 (GNU_HASH) 0x210
  0x0000000000000000 (NULL)     0x0
```

File: lldb/test/Shell/ObjectFile/ELF/elf-dynsym.test [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// HASH: Dumping headers for 1 module(s).
+// HASH: ObjectFileELF, file =
+// HASH: ELF Header
+// HASH: e_type      = 0x0003 ET_DYN
+// Make sure there are no section headers
+// HASH: e_shnum = 0x00000000
+// Make sure we were able to load the symbols
+// HASH: Symtab, file = {{.*}}elf-dynsym.test.tmp.sysv, num_symbols = 2:
+// HASH-DAG: undefined
+// HASH-DAG: defined
```
Comment (labath) [inline lldb/test/Shell/ObjectFile/ELF/elf-dynsym.test:42]: In cases like this where the two outputs are (nearly) identical, I think its better to have a one check prefix for the common parts of the output -- it's less code and it better emphasizes the difference. This can be achieved by passing different sets of prefixes to the FileCheck command:
```
// RUN: ... | FileCheck --check-prefixes=BOTH,GNU
// RUN: ... | FileCheck --check-prefixes=BOTH,SYSV
// BOTH: This line is common for both formats
// GNU: This is gnu-specific output
// SYSV: This is sysv-specific output
```

**Comment by rastogishubham** [comment]
It seems like this patch is responsible for breaking greendragon 

https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15470/

**Comment by clayborg** [comment]
> I have reverted this change with [f14e1a8](https://github.com/llvm/llvm-project/commit/f14e1a8597f83fa5bbc78befcb7059144d58ff5c)

I found the issue and have a fix.
